python/tk_3dsmaxplus/maxscript.py

```diff
@@ -91,9 +91,30 @@ def add_action_to_menu(callback, action_name, menu_var, engine):
         obj = callback.im_self
         method_name = callback.__name__
 
-        # Need a globally available version of this object for maxscript action callbacks to be able to refer to python objects
-        object_id = str(id(obj))
-        engine.maxscript_objects[object_id] = obj
+        # Note that we're using the action name because we need these
+        # macros to reference things consistently across sessions. Sadly,
+        # if a second, concurrent, 3ds Max session is launched, Toolkit
+        # will build the Shotgun menu in that session and Max will write
+        # that updated menu layout to disk for the user, because it thinks
+        # that needs to persist across sessions. This causes us problems
+        # in the first session, then, because Max looks up what macro to
+        # run from the xml stored on disk when the action is triggered.
+        # This means that if we have anything referenced from the macro
+        # that is not available in the first session, the action will
+        # fail.
+        hash_name = action_name
+
+        # This won't be visible to the user, so we'll go the ugly route
+        # to resolve conflicts and just append underscores until we get
+        # a unique key. Ultimately, this is just covering what should be
+        # the very rare edge case of having two menu actions with the
+        # same name. That would be bad practice, in my opinion, but since
+        # it is possible we will handle it.
+        while hash_name in engine.maxscript_objects:
```

Comment:
It can't be anything unique to the session, which includes time and the macro name, which might change across sessions. Technically even what I'm doing might end up with a mismatch between sessions if the two share some menu actions and not others (and includes a duplicate action name in one or both), but it's an exceedingly unlikely edge case that I'm willing to accept.